src/main/java/com/amazon/aws/iot/greengrass/semver/SemVer.java

```diff
@@ -0,0 +1,340 @@
+        int maj = compareIdentifiers(this.major, other.major);
+        if (maj != 0) {
+            return maj;
+        }
+        int min = compareIdentifiers(this.minor, other.minor);
+        if (min != 0) {
+            return min;
+        }
+        return compareIdentifiers(this.patch, other.patch);
+    }
+
+    public int comparePre(SemVer other) {
+        // NOT having a prerelease is > having one
+        if (!this.prerelease.isEmpty() && other.prerelease.isEmpty()) {
+            return -1;
+        } else if (this.prerelease.isEmpty() && !other.prerelease.isEmpty()) {
+            return 1;
+        } else if (this.prerelease.isEmpty() && other.prerelease.isEmpty()) {
+            return 0;
+        }
+
+        for (int i = 0; ; i++) {
+            if (i > this.prerelease.size() - 1 && i > other.prerelease.size() - 1) {
+                return 0;
+            } else if (i > other.prerelease.size() - 1) {
+                return 1;
+            } else if (i > this.prerelease.size() - 1) {
+                return -1;
+            }
+            Object a = this.prerelease.get(i);
+            Object b = other.prerelease.get(i);
+            if (Objects.equals(a, b)) {
+                continue;
+            }
+            return compareIdentifiers(Objects.toString(a), Objects.toString(b));
+        }
+    }
+
+    public int compareBuild(SemVer other) {
+        for (int i = 0; ; i++) {
+            if (i > this.build.size() - 1 && i > other.build.size() - 1) {
+                return 0;
+            } else if (i > other.build.size() - 1) {
+                return 1;
+            } else if (i > this.build.size() - 1) {
+                return -1;
+            }
+            String a = this.build.get(i);
+            String b = other.build.get(i);
+            if (Objects.equals(a, b)) {
+                continue;
+            } else {
+                return compareIdentifiers(a, b);
+            }
+        }
+    }
+
+    static Integer isInt(String x) {
+        try {
+            return Integer.parseInt(x);
+        } catch (NumberFormatException f) {
+            return null;
+        }
+    }
+
+    public static int compareIdentifiers(int a, int b) {
+        return Integer.compare(a, b);
+    }
+
+    public static int compareIdentifiers(String a, String b) {
+        Integer anum = isInt(a);
+        Integer bnum = isInt(b);
+
+        int comparison = a.equals(b) ? 0 : (anum != null && bnum == null) ? -1 : (bnum != null && anum == null) ? 1
+                : anum != null && bnum != null ? (anum < bnum ? -1 : 1) : a.compareTo(b);
+        if (comparison >= 1) {
+            return 1;
+        } else if (comparison == 0) {
+            return 0;
+        }
+        return -1;
+    }
+
+    // preminor will bump the version up to the next minor release, and immediately
+    // down to pre-release. premajor and prepatch work the same way.
+    public SemVer inc(String release, String identifier) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.